clients/drcachesim/tests/cache_replacement_policy_unit_test.cpp

```diff
@@ -0,0 +1,133 @@
+            exit(1);
+        }
+    }
+
+    void
+    unit_test_replace_which_way()
+    {
+        // Create and initialize a 4-way set associative cache with line size of 32 and
+        // total size of 256 bytes.
+        initialize_cache(4, 32, 256);
+
+        memref_t ref_1;
+        ref_1.data.type = TRACE_TYPE_READ;
+        ref_1.data.size = 1;
+        // Access the cache line in the following fashion. This sequence follows the
+        // sequence shown in https://github.com/DynamoRIO/dynamorio/issues/4881.
+        // Access the first row.
+        ref_1.data.addr = 0;
+        request(ref_1); // This accesses "a" in issue 4881.
+
+        ref_1.data.addr = 64;
+        request(ref_1); // This accesses "b" in issue 4881.
+
+        ref_1.data.addr = 128;
+        request(ref_1); // This accesses "c" in issue 4881.
+
+        ref_1.data.addr = 192;
+        request(ref_1); // This accesses "d" in issue 4881.
+
+        // After the above accesses, the counters for each way will be as follows:
+        //  way 0 ("a" in issue 4881): 3
+        //  way 1 ("b" in issue 4881): 2
+        //  way 2 ("c" in issue 4881): 1
+        //  way 3 ("d" in issue 4881): 0
+        // At this point way 0 ("a") has the highest counter so it should be replaced by
+        // the LRU policy.
+        assert(replace_which_way(0) == 0);
+
+        ref_1.data.addr = 0;
+        request(ref_1); // This replaces way 0 ("a").
+        // At this point way 0 has been replaced(accessed) and way 1 has the highest
+        // counter.
+        //  way 0 ("a" in issue 4881): 0
+        //  way 1 ("b" in issue 4881): 3
+        //  way 2 ("c" in issue 4881): 2
+        //  way 3 ("d" in issue 4881): 1
+        assert(replace_which_way(0) == 1);
+
+        ref_1.data.addr = 64;
+        request(ref_1); // This replaces way 1 ("b").
+        // At this point way 1 has been replaced(accessed) and way 2 has the highest
+        // counter.
+        //  way 0 ("a" in issue 4881): 1
+        //  way 1 ("b" in issue 4881): 0
+        //  way 2 ("c" in issue 4881): 3
+        //  way 3 ("d" in issue 4881): 2
+        assert(replace_which_way(0) == 2);
+
+        ref_1.data.addr = 128;
+        request(ref_1); // This replaces way 2 ("c").
+        // At this point way 2 has been replaced(accessed) and way 3 has the highest
+        // counter.
+        //  way 0 ("a" in issue 4881): 2
+        //  way 1 ("b" in issue 4881): 1
+        //  way 2 ("c" in issue 4881): 0
+        //  way 3 ("d" in issue 4881): 3
+        assert(replace_which_way(0) == 3);
+
+        ref_1.data.addr = 192;
+        request(ref_1); // This replaces way 3 ("d").
+        // At this point way 3 has been replaced(accessed) and way 0 has the highest
+        // counter.
+        //  way 0 ("a" in issue 4881): 3
+        //  way 1 ("b" in issue 4881): 2
+        //  way 2 ("c" in issue 4881): 1
+        //  way 3 ("d" in issue 4881): 0
+        assert(replace_which_way(0) == 0);
+    }
```

Comment:
More test scenarios would be good.  The above does not actually test replacing anything: so testing that the right thing happens when new addresses come in and evict prior data would be good, e.g.  Other corner cases you can think of would be good too.  As separate scenarios with a new cache might be best.

Or if you prefer we could commit just the #4881 sequence first and add more separately, but please put in a `XXX i#4842: Add more test sequences` comment since otherwise the reader will look at this and assume that the author thinks this is complete.

Reply:
Added the comment to add more sequences in subsequent commits.